### ISSUE TITLE
Pattern-match common decomposed ops in the executable (softmax, DPA, ...)

### ISSUE THREAD

**Issue by tillahoffmann** [tillahoffmann/jax-mps#133]
## Motivation

Several high-value fused MLX kernels correspond to JAX APIs that are Python-composed (no single primitive to intercept at trace time):

- \`jax.nn.softmax\` / \`log_softmax\` / \`jax.scipy.special.logsumexp\` → \`mlx::core::softmax\` / \`logsumexp\`
- \`jax.nn.dot_product_attention\` (user-written or via \`jax.nn\`) → \`mlx::core::fast::scaled_dot_product_attention\`
- Hand-rolled attention: \`softmax(Q @ K^T / sqrt(d)) @ V\` — same SDPA kernel applies

\`register_lowering\` doesn't work here (no single primitive). Explicit new primitives (\`mps.softmax\`, \`mps.sdpa\`) exist but require users to opt in.

## Proposal

Detect these patterns in the lowered StableHLO before dispatch, rewrite them to a single \`stablehlo.custom_call @mps.<fused_name>\`, then let the existing handler execute the fused MLX kernel. Zero user code change; catches in-the-wild code that didn't know about our explicit primitives.

### softmax family

\`\`\`
%m = reduce_max(x, axis)
%s = sub(x, broadcast(m))
%e = exp(s)
%z = reduce_sum(e, axis)
%out = divide(e, broadcast(z))   ← softmax
%out = sub(s, log(broadcast(z))) ← log_softmax
%out = add(m, log(z))            ← logsumexp
\`\`\`

### scaled dot-product attention

\`\`\`
%qk = dot_general(Q, K) {contracting=[last], ...}
%scaled = divide(qk, sqrt(d))          # or multiply by 1/sqrt(d)
[optional mask: add / select]
%attn = <softmax pattern above>
%out = dot_general(attn, V) {contracting=[last of attn, first of V], ...}
\`\`\`

Two masked variants to distinguish: additive mask (\`add\`) vs boolean mask (\`select\`); causal is a triangular constant mask that can be recognized and routed to \`mps.sdpa_causal\`.

## Where this lives

Likely a StableHLO rewrite pass applied in the \`StablehloToMlxExecutable\` pipeline before the per-op dispatch loop. Similar in shape to the existing optimization passes.

## Risks & costs

- **Fragility**: JAX can change its decomposition at any version bump. Need a regression-test matrix asserting each target pattern still rewrites. Tie to CI across supported JAX versions.
- **False positives**: a hand-rolled \`reduce_max → sub → exp → reduce_sum → div\` that's NOT a softmax (e.g., stopped-gradient variant, custom normalization) could be rewritten incorrectly. Be conservative — only match the exact shape, document tolerances.
- **Maintenance**: each new pattern is ongoing code to own.

## Related

- #125 — \`dot_product_attention\` interception via \`register_lowering\` (alternative route)
- sibling ticket — explicit \`mps.softmax\` primitives (opt-in route)

Three routes to the same fused kernels. This issue is the "no user code change" route, which is the most ambitious and most maintenance-heavy.
